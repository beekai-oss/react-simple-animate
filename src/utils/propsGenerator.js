// @flow
import type { State, Props } from '../animate';

let style;
let transition;

type MountProps = { willUnmount: boolean, willMount: boolean };

const getTransitionDelay = ({ reverseDelaySeconds, startAnimation, played, delaySeconds }) => {
  if (reverseDelaySeconds && !startAnimation && played) {
    return ` ${reverseDelaySeconds}s`;
  } else if (delaySeconds) {
    return ` ${delaySeconds}s`;
  }
  return '';
};

export default function propsGenerator(
  {
    startAnimation,
    startStyle,
    endStyle,
    onCompleteStyle,
    durationSeconds = 0.3,
    reverseDelaySeconds,
    delaySeconds = 0,
    easeType = 'linear',
    className,
    transition: transitionValue,
    refCallback,
  }: Props,
  { willComplete, willEnter, willLeave, played }: State,
  mountProps: MountProps = {
    willUnmount: false,
    willMount: false,
  },
) {
  const { willUnmount, willMount } = mountProps;
  style = startStyle;
  transition =
    transitionValue ||
    `${durationSeconds}s all ${easeType}${getTransitionDelay({ reverseDelaySeconds, startAnimation, played, delaySeconds })}`;

  if (willMount) {
    style = willEnter ? endStyle : startStyle;
  } else if (willUnmount) {
    style = willLeave ? endStyle : startStyle;
  } else if (willComplete && onCompleteStyle) {
    style = onCompleteStyle;
    transition = null;
  } else if (startAnimation) {
    style = endStyle;
  }

  return {
    className,
    style: {
      ...{
        ...style,
        transition,
      },
    },
    ...(refCallback
      ? {
        ref: (element: any): void => {
          if (element && refCallback) refCallback(element);
        },
      }
      : null),
  };
}
