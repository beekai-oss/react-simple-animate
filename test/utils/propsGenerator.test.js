import propsGenerator from '../../src/utils/propsGenerator';

describe('propsGenerator', () => {
  const state = {
    willEnd: false,
    willStart: false,
    willComplete: false,
    willEnter: false,
    willLeave: false,
    played: false,
  };

  const props = {
    startStyle: {
      display: 'none',
    },
    endStyle: {
      display: 'block',
    },
    startAnimation: false,
  };

  const startStyleWithTransition = {
    ...props.startStyle,
    transition: '0.3s all linear',
  };

  const endStyleWithTransition = {
    ...props.endStyle,
    transition: '0.3s all linear',
  };

  it('should generate startStyle', () => {
    expect(propsGenerator(props, state).style).toEqual(
      startStyleWithTransition,
    );
  });

  describe('when animation is set to true', () => {
    it('should generate endStyle', () => {
      expect(
        propsGenerator({ ...props, startAnimation: true }, state).style,
      ).toEqual(endStyleWithTransition);
    });
  });

  describe('when item about to mount', () => {
    it('should render the start style', () => {
      expect(
        propsGenerator({ ...props, startAnimation: true }, state, {
          willMount: true,
        }).style,
      ).toEqual(startStyleWithTransition);
    });

    it('should render the end style when state set animation about to enter', () => {
      expect(
        propsGenerator(
          { ...props, startAnimation: true },
          { ...state, willEnter: true },
          { willMount: true },
        ).style,
      ).toEqual(endStyleWithTransition);
    });
  });

  describe('when item about to unmount', () => {
    it('should render the start style', () => {
      expect(
        propsGenerator({ ...props, startAnimation: true }, state, {
          willUnmount: true,
        }).style,
      ).toEqual(startStyleWithTransition);
    });

    it('should render the end style when state set animation about to leave', () => {
      expect(
        propsGenerator(
          { ...props, startAnimation: true },
          { ...state, willLeave: true },
          { willUnmount: true },
        ).style,
      ).toEqual(endStyleWithTransition);
    });
  });

  describe('when animation reverseDelaySeconds has been set', () => {
    it('should render the end style', () => {
      expect(
        propsGenerator(
          { ...props, startAnimation: true, reverseDelaySeconds: 0.3 },
          state,
        ).style,
      ).toEqual(endStyleWithTransition);
    });

    it('should render the end style when state set animation about to leave', () => {
      expect(
        propsGenerator(
          { ...props, startAnimation: false, reverseDelaySeconds: 0.3 },
          { ...state, willStart: true },
        ).style,
      ).toEqual(startStyleWithTransition);
    });
  });

  it('should render end style when willEnd is set to true', () => {
    expect(
      propsGenerator(
        { ...props, startAnimation: true },
        { ...state, willEnd: true },
      ).style,
    ).toEqual(endStyleWithTransition);
  });

  it('should render start style when delay second is set and animation about to start', () => {
    expect(
      propsGenerator(
        { ...props, startAnimation: true, delaySeconds: 0.3 },
        state,
      ).style,
    ).toEqual({
      ...endStyleWithTransition,
      transition: '0.3s all linear 0.3s',
    });
  });

  it('should render complete style and set transition to null', () => {
    expect(
      propsGenerator(
        {
          ...props,
          onCompleteStyle: {
            display: 'flex',
          },
        },
        { ...state, willComplete: true },
      ).style,
    ).toEqual({ display: 'flex', transition: null });
  });
});
