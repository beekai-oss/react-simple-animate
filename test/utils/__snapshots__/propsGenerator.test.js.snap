// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`propsGenerator should set up refCallback if it is part of the prop 1`] = `
Object {
  "className": undefined,
  "ref": [Function],
  "style": Object {
    "transition": "all 0.3s linear 0s",
  },
}
`;
